internal/engine/up_watch.go

```diff
@@ -0,0 +1,254 @@
+	go func() {
+		defer close(ret)
+		for {
+			event, ok := <-eventChan
+			if !ok {
+				return
+			}
+			events := []fsnotify.Event{event}
+
+			// keep grabbing changes until we've gone `watchBufferMinRestDuration` without seeing a change
+			minRestTimer := timerMaker(watchBufferMinRestDuration)
+
+			// but if we go too long before seeing a break (e.g., a process is constantly writing logs to that dir)
+			// then just send what we've got
+			timeout := timerMaker(watchBufferMaxDuration)
+
+			done := false
+			channelClosed := false
+			for !done && !channelClosed {
+				select {
+				case event, ok := <-eventChan:
+					if !ok {
+						channelClosed = true
+					} else {
+						minRestTimer = timerMaker(watchBufferMinRestDuration)
+						events = append(events, event)
+					}
+				case <-minRestTimer:
+					done = true
+				case <-timeout:
+					done = true
+				}
+			}
+			if len(events) > 0 {
+				ret <- events
+			}
+
+			if channelClosed {
+				return
+			}
+		}
+
+	}()
+	return ret
+}
+
+func addEventsToLeftover(leftover map[string]*serviceFilesChangedEvent, events []serviceSingleFileChangeEvent) {
+	for _, event := range events {
+		if _, exists := leftover[string(event.service.Name)]; !exists {
+			leftover[string(event.service.Name)] = &serviceFilesChangedEvent{event.service, []string{event.fileName}}
+		} else {
+			leftover[string(event.service.Name)].files = append(leftover[string(event.service.Name)].files, event.fileName)
+		}
+	}
+}
+
+type watchEventsStream struct {
+	events <-chan []serviceSingleFileChangeEvent
+	errs   <-chan error
+}
+
+type serviceSingleFileChangeEvent struct {
+	service  model.Service
+	fileName string
+}
+
+type serviceNotifyPair struct {
+	service model.Service
+	notify  watch.Notify
+}
+
+func makeFilter(ctx context.Context, service model.Service) (git.IgnoreTester, error) {
+	var repoRoots []string
+
+	for _, mount := range service.Mounts {
+		repoRoots = append(repoRoots, mount.Repo.LocalPath)
+	}
+
+	eventFilter, err := git.NewMultiRepoIgnoreTester(ctx, repoRoots)
+	if err != nil {
+		return nil, err
+	}
+
+	return eventFilter, nil
+}
+
+// turns a list of (service, chan fsevent) pairs into a single chan (service, fsevent)
+func snsToServiceWatcher(ctx context.Context, timerMaker timerMaker, sns []serviceNotifyPair) (*watchEventsStream, error) {
+	events := make(chan []serviceSingleFileChangeEvent)
+	errs := make(chan error)
+
+	for _, sn := range sns {
+		coalescedEvents := coalesceEvents(timerMaker, sn.notify.Events())
+		filter, err := makeFilter(ctx, sn.service)
+		if err != nil {
+			return nil, err
+		}
+
+		go func(service model.Service, watcher watch.Notify) {
+			for {
+				select {
+				case err, ok := <-watcher.Errors():
+					if !ok {
+						close(events)
+						close(errs)
+						return
+					}
+					errs <- err
+				case fsEvents, ok := <-coalescedEvents:
+					if !ok {
+						close(events)
+						close(errs)
+						return
+					}
+					var watchEvents []serviceSingleFileChangeEvent
```

Comment:
i think this code will be simpler if you separate out the filtering from the multiplexing. merging multiple channels is kind of tricky. filtering/mapping is easy, because you can use a range on the input channel

for event := range myChannel {
  ...
  newChannel <- newEvent
}

Reply:
per offline discussion, this starts getting uglier since we need to handle errors, and we can figure it out later